### PR TITLE
roachtest: add `failover/non-system/crash`

### PR DESCRIPTION
**Pull request by erikgrinaker**
`failover/non-system/crash` benchmarks the maximum duration of range
unavailability following a leaseholder crash with only non-system
ranges. It tests the simplest possible failure:

  - A process crash, where the host/OS remains available (in particular, the
    TCP/IP stack is responsive and sends immediate RST packets to peers).

  - No system ranges located on the crashed node.

  - SQL clients do not connect to the crashed node.

  - The workload consists of individual point reads and writes.

Since the lease unavailability is probabilistic, depending e.g. on the
time since the last heartbeat and other variables, we run 9 crashes and
record the pMax latency to find the upper bound on unavailability. We
expect this worse-case latency to be slightly larger than the lease
interval (9s), to account for lease acquisition and retry latencies. We
do not assert this, but instead export latency histograms for graphing.

The cluster layout is as follows:

n1-n3: System ranges and SQL gateways.
n4-n6: Workload ranges.
n7:    Workload runner.

The test runs a kv50 workload with batch size 1, using 256 concurrent
workers directed at n1-n3 with a rate of 2048 reqs/s. n4-n6 are killed
and restarted in order, with 30 seconds between each operation, for 3
cycles totaling 9 crashes.

Touches #79494.
Epic: CRDB-18520.

Release note: None

### CODE REVIEW

**Comment by erikgrinaker** [comment]
Been wondering if we should maybe extend `workload` to discard latencies below e.g. 1 second, such that we can get p50 and p99 latencies _during leaseholder loss_ instead of over the entire period (which is drowned out by nominal requests between the crashes). Wdyt?

**Comment by tbg** [comment]
Don't workload's `1s` intervals (`stats.json`)give you what you want? At least if you run with high concurrency and a rate limit (to simulate an open-loop workload). Adding filters on what to record to workload histograms seems too bespoke.
Also, progress will basically halt during the leaseholder loss, no? So aren't we going to record nothing for ~9s, then a bunch of 9s latencies?

**Comment by erikgrinaker** [comment]
I'm interested in the p50 and pMax across all crashes. The unavailability will vary depending on the timing of the crash relative to the start of the lease. But yeah, I should be able to get that from the stats histograms -- I've just been using the final `workload` output while trying this stuff out, whose p50 value is useless.

**Review by tbg — COMMENTED** [review]
Why is a p50 (or even pmax) interesting across the crash? Just to guess at the proportion of requests that hit a temporary unavailability? Seems like a confusing measure to look at.

The p100 makes more sense, as it could be understood as simply measuring the duration of the outage. I think it's a reasonable thing to be looking at it, but there are pitfalls too. If some requests get stuck forever, they'll never be added to the histogram and so the p100 might look better than it actually is.

Perhaps rather than using a workload here, we could leverage `crdb_internal.probe_ranges`? We'd introduce the outage and immediately start a probe (with a large enough timeout), and then the latency of that is the duration of the outage. If the probe doesn't succeed within the timeout, it will report per-range failures and we can record that as an "infinity" measurement.

**Comment by erikgrinaker** [comment]
>Why is a p50 (or even pmax) interesting across the crash? Just to guess at the proportion of requests that hit a temporary unavailability? Seems like a confusing measure to look at.
>
>The p100 makes more sense, as it could be understood as simply measuring the duration of the outage. I think it's a reasonable thing to be looking at it, but there are pitfalls too. If some requests get stuck forever, they'll never be added to the histogram and so the p100 might look better than it actually is.

If there was only a single outage, then the pMax would be sufficient. We're sampling 9 outages here, and I want to know the median (p50) outage duration as well as the worst-case outage duration, since the median is a better measure of what to typically expect in the wild. These durations are probabilistic across a wide timespan, as seen in #91947.

I suppose it would be better to segment the data into discrete outages and measure the pMax for each individual outage, then take the median of that. I was hoping to avoid doing anything fancy here and reuse existing functionality in workload/roachperf, but it shouldn't be a huge lift. I'll do the data analysis and graphing later.

> Perhaps rather than using a workload here, we could leverage `crdb_internal.probe_ranges`? We'd introduce the outage and immediately start a probe (with a large enough timeout), and then the latency of that is the duration of the outage. If the probe doesn't succeed within the timeout, it will report per-range failures and we can record that as an "infinity" measurement.

I want this to approximate real-world conditions, since I want to know what users can expect in the wild (except for the test constraints, which are useful to separate individual failure modes). This will be more relevant for later tests that introduce other faults and workloads -- in particular, I also want to look at a scan-like workload that touches ranges sequentially in the face of network outages, where each range touched currently would have to wait for a sequential network timeout.

File: pkg/cmd/roachtest/tests/lease.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+// n1-n3: System ranges and SQL gateways.
+// n4-n6: Workload ranges.
+// n7:    Workload runner.
+//
+// The test runs a kv50 workload with batch size 1, using 256 concurrent
+// workers directed at n1-n3. n4-n6 are killed and restarted in order, with 30
+// seconds between each operation, for 3 cycles totaling 9 crashes.
+func runAcquireLeaseCrashNonSystem(ctx context.Context, t test.Test, c cluster.Cluster) {
+	c.Put(ctx, t.Cockroach(), "./cockroach")
+	c.Put(ctx, t.DeprecatedWorkload(), "./workload")
```
Comment (tbg) [inline pkg/cmd/roachtest/tests/lease.go:66]: Why? `./cockroach workload` not enough?

File: pkg/cmd/roachtest/tests/lease.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+	// Create cluster with one rack locality per node and configure it.
+	require.Equal(t, 7, c.Spec().NodeCount)
+
+	opts := option.DefaultStartOpts()
+	settings := install.MakeClusterSettings(install.NumRacksOption(c.Spec().NodeCount))
```
Comment (tbg) [inline pkg/cmd/roachtest/tests/lease.go:72]: `roachprod` clusters come with some attributes included, so you shouldn't need to bother with this.

https://github.com/cockroachdb/cockroach/blob/ae0ad5ca29ff865324cfc52bf75c75357f8af592/pkg/roachprod/install/cockroach.go#L581

File: pkg/cmd/roachtest/tests/lease.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+		_, err = conn2.ExecContext(ctx, fmt.Sprintf(
+			`ALTER %s CONFIGURE ZONE USING num_replicas = 3, constraints = '[-rack=3, -rack=4, -rack=5]'`,
+			target))
+		require.NoError(t, err)
+	}
+	require.NoError(t, rows.Err())
+	rows.Close()
+	require.NoError(t, conn2.Close())
+
+	// Wait for upreplication.
```
Comment (tbg) [inline pkg/cmd/roachtest/tests/lease.go:97]: Note that this doesn't wait for the 5x replication expected for system ranges here.

File: pkg/cmd/roachtest/tests/lease.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+	t.Status("creating workload database")
+	_, err = conn.ExecContext(ctx, `CREATE DATABASE kv`)
+	require.NoError(t, err)
+	_, err = conn.ExecContext(ctx, `ALTER DATABASE kv CONFIGURE ZONE USING `+
+		`num_replicas = 3, constraints = '[-rack=0, -rack=1, -rack=2]'`)
+	require.NoError(t, err)
+	c.Run(ctx, c.Node(7), `./workload init kv --splits 1000 {pgurl:1}`)
+
+	// The replicate queue takes forever to move the kv ranges from n1-n3 to
+	// n4-n6, so we do it ourselves.
```
Comment (tbg) [inline pkg/cmd/roachtest/tests/lease.go:111]: Wouldn't it be faster to do this first, before creating 1000 ranges?

File: pkg/cmd/roachtest/tests/lease.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+	for rows.Next() {
+		var target string
+		require.NoError(t, rows.Scan(&target))
+		_, err = conn2.ExecContext(ctx, fmt.Sprintf(
+			`ALTER %s CONFIGURE ZONE USING num_replicas = 3, constraints = '[-rack=3, -rack=4, -rack=5]'`,
+			target))
+		require.NoError(t, err)
+	}
+	require.NoError(t, rows.Err())
+	rows.Close()
```
Comment (tbg) [inline pkg/cmd/roachtest/tests/lease.go:94]: nit: this call is not necessary.

File: pkg/cmd/roachtest/tests/lease.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	c.Start(ctx, t.L(), opts, settings, c.Range(1, 6))
+
+	conn := c.Conn(ctx, t.L(), 1)
+
+	t.Status("configuring cluster")
+	_, err := conn.ExecContext(ctx, `SET CLUSTER SETTING kv.range_split.by_load_enabled = 'false'`)
+	require.NoError(t, err)
+
+	// Constrain all existing zone configs to n1-n3.
+	conn2 := c.Conn(ctx, t.L(), 1)
```
Comment (tbg) [inline pkg/cmd/roachtest/tests/lease.go:82]: Why another conn?

File: pkg/cmd/roachtest/tests/lease.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+				`WHERE database_name = 'kv' AND $1::int = ANY(replicas)`, source).Scan(&count))
+			if count == 0 {
+				break
+			}
+			t.Status(fmt.Sprintf("moving %d kv ranges off of n%d", count, source))
+			for _, target := range []int{4, 5, 6} {
+				_, err = conn.ExecContext(ctx, `ALTER RANGE RELOCATE FROM $1::int TO $2::int FOR `+
+					`SELECT range_id FROM crdb_internal.ranges WHERE database_name = 'kv' AND $1::int = ANY(replicas)`,
+					source, target)
+				require.NoError(t, err)
```
Comment (tbg) [inline pkg/cmd/roachtest/tests/lease.go:125]: Isn't this just gonna flake? Or is `ALTER RANGE RELOCATE` guaranteed to retry all errors away internally?

**Review by tbg — COMMENTED** [review]
Collecting meaningful data from this roachtest is somewhat challenging (see earlier discussion on this PR).

Having the workload run across multiple outages seems to make it more challenging to interpret the results. I'm curious why you opted for multiple outages. Is this just to amortize the overhead of setting up the roachtest?

I also forget what roachperf can handle. Does it even support multiple workloads (multiple stats.json files)? If it doesn't, then adding more workloads here will be challenging (ideally we don't have to have one roachtest per workload).

I think what you have here works, you could read the `stats.json` files and segment the files by timestamp, then produce the per-event percentiles you want (earlier it sounded like you wanted the p50 of requests, now it sounds like you want the p50 of the per-event p100s which seems plausible). I seem to remember that we have code that can read the stats.json files, but I might be wrong (and they're just JSON anyway so not hard to parse).

The clusterstats stuff can then report the results to roachdash: https://github.com/cockroachdb/cockroach/blob/bb23d2aa1a69132fa86ac5315e94f7b13d4f4ce6/pkg/cmd/roachtest/clusterstats/doc.go

The earlier point about a statement timeout still stands. I'm not sure I trust the statement timeout to actually be strictly enforced, I've seen commands where it wasn't, plus not sure we can get it into the workload. `workload`-level enforcement would be better, probably the same amount of work anyway but if statement stuff is much easier I'd be ok with it for now too.

Overall looks good! Will be interesting to see the results of this.

**Comment by erikgrinaker** [comment]
>Having the workload run across multiple outages seems to make it more challenging to interpret the results. I'm curious why you opted for multiple outages. Is this just to amortize the overhead of setting up the roachtest?

The lease acquisition latency is very variable, from 6 to 11 seconds. I'm interested in finding the worst-case acquisition latency and graph that. If we only sampled a single crash, the graph would be extremely noisy and essentially useless. By sampling 9 crashes, we approach the real worst-case lease acquisition latency.

The p50 lease acquisition latency is secondary -- if it's trivial to include then let's do it (it's also useful as an expected value), but I won't jump through a bunch of hoops for it.

>I also forget what roachperf can handle. Does it even support multiple workloads (multiple stats.json files)? If it doesn't, then adding more workloads here will be challenging (ideally we don't have to have one roachtest per workload).

I'm planning to add separate roachtest variants for separate workloads. We could reuse some code if we want to. But I think it's useful to keep separate latency measurements for separate kinds of scenarios, since it may require separate mechanisms to properly address them.

File: pkg/cmd/roachtest/tests/lease.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/require"
+)
+
+func registerLease(r registry.Registry) {
```
[the remark this answers is not in the document]
Comment (erikgrinaker) [reply pkg/cmd/roachtest/tests/lease.go:28]: Good point, failover makes sense. We'll also end up testing other stuff here, like Raft elections and network timeouts.

File: pkg/cmd/roachtest/tests/lease.go (hunk 1 above)
Comment (erikgrinaker) [reply pkg/cmd/roachtest/tests/lease.go:66]: Just copy/pasted this. It does seem like`workload` has more functionality than `cockroach workload` though, and I always end up reaching for it. 🤷‍♂️

File: pkg/cmd/roachtest/tests/lease.go (hunk 3 above)
Comment (erikgrinaker) [reply pkg/cmd/roachtest/tests/lease.go:97]: We explicitly set them to 3x replication above. We don't want them on the workload nodes.

File: pkg/cmd/roachtest/tests/lease.go (hunk 4 above)
Comment (erikgrinaker) [reply pkg/cmd/roachtest/tests/lease.go:111]: You would think so, but `workload init kv --splits` command sprays the ranges across all nodes regardless of the initial range layout. Tried a bunch of stuff to make it position them correctly without any range movement, and didn't want to start messing around with patching `workload`.

File: pkg/cmd/roachtest/tests/lease.go (hunk 7 above)
Comment (erikgrinaker) [reply pkg/cmd/roachtest/tests/lease.go:125]: It outputs errors as strings in the result table.

File: pkg/cmd/roachtest/tests/lease.go (hunk 6 above)
Comment (erikgrinaker) [reply pkg/cmd/roachtest/tests/lease.go:82]: Because we're reading rows from `conn` in the loop. I could have buffered the results in a slice, but this has the same effect with less code.

File: pkg/cmd/roachtest/tests/lease.go (hunk 1 above)
Comment (erikgrinaker) [reply pkg/cmd/roachtest/tests/lease.go]: Updated to use `cockroach workload`.

File: pkg/cmd/roachtest/tests/lease.go (hunk 2 above)
Comment (erikgrinaker) [reply pkg/cmd/roachtest/tests/lease.go]: Nice, thanks. Updated.

**Review by tbg — APPROVED** [review]
:lgtm:

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91810#-:-NHxtIfkBoBjIIKfi7Fx:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker, @herkolategan, and @smg260)

___
*[`pkg/cmd/roachtest/tests/failover.go` line 81 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91810#-NHxrsPb-dauCOV8736N:-NHxrsPb-dauCOV8736O:b-wlspu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/168dcd8b31f4ba5fec94e4404529b56e3df375b7/pkg/cmd/roachtest/tests/failover.go#L81)):*
> This test controls the ranges manually.
___
*[`pkg/cmd/roachtest/tests/lease.go` line 82 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91810#-NGmRhLPp4_GNmX0gUsm-r1-82:-NHxsH581AtigwJ5kiBf:bgnmoep) ([raw file](https://github.com/cockroachdb/cockroach/blob/5643217932e052e52e3a5dd690ec7203cf1dfafc/pkg/cmd/roachtest/tests/lease.go#L82)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Because we're reading rows from `conn` in the loop. I could have buffered the results in a slice, but this has the same effect with less code.
</blockquote></details>

`c.Conn` is a `*gosql.DB` which is a connection pool, so the fixup commit I just pushed should work - might be missing something?
```
```

**Comment by erikgrinaker** [comment]
CI failures appear unrelated, giving this a go.

bors r+

**Comment by erikgrinaker** [comment]
Rebased.

bors r+
